contributors/design-proposals/storage/csi-migration.md

```diff
@@ -0,0 +1,443 @@
+* Framework for plugin migration built for Dynamic provisioning, pre-provisioned
+  volumes, and in-tree volumes
+
+Beta [Target 1.14]
+* On by default
+* Migrate all of the cloud provider plugins*
+
+GA [TBD]
+* Feature on by default, per-plugin toggle on for relevant cloud provider by
+  default
+* CSI Drivers for migrated plugins available on related cloud provider cluster
+  by default
+
+## Feature Gating
+We will have an alpha feature gate for the whole feature that can turn the CSI
+migration on or off, when off all code paths should revert/stay with the in-tree
+plugins. We will also have individual flags for each driver so that admins can
+toggle them on or off.
+
+The feature gate can exist at the interception points in the OperationGenerator
+for Attach and Mount, as well as in the PV Controller for Provisioning.
+
+We will also have one feature flag for each driver’s migration so that each
+driver migration can be turned on and off individually. 
+
+The new feature gates for alpha are:
+```
+// Enables the in-tree storage to CSI Plugin migration feature.
+CSIMigration utilfeature.Feature = "CSIMigration"
+
+// Enables the GCE PD in-tree driver to GCE CSI Driver migration feature.
+CSIMigrationGCE utilfeature.Feature = "CSIMigrationGCE"
+
+// Enables the AWS in-tree driver to AWS CSI Driver migration feature.
+CSIMigrationAWS utilfeature.Feature = "CSIMigrationAWS"
+```
+
+## Translation Layer
+The main mechanism we will use to migrate plugins is translating In-tree plugin
+sources into CSI Sources and redirecting operation calls to the CSI Driver
+instead of the in-tree driver.
+
+Pros:
+* Keeps old API objects as they are
+* Facilitates gradual roll-over to CSI
+
+Cons:
+* Somewhat complicated and error prone.
+* AttachDetach Controller must have permissions to modify PVs.
+* Bespoke translation logic for each in-tree plugin
+
+### Dynamically Provisioned Volumes
+
+#### Kubernetes Changes
+In-tree provisioner will be changed to have a feature gate that when turned on
+will not register the in-tree volume plugin into PV controller. We will modify
+`provisionClaimOperation` to set the claim provisioner to the migrated external
+provisioner name, so the external provisioner will pick this up and treat it as
+a CSI Volume from then on. The controller will automatically annotate PVCs for
+external provisioning by an provisioner that handles e.g. “google.io/gce-pd”.
+
+#### External Component Changes
+External provisioner will use a different  provisioner name as the internal
+volume plugin name (e.g. “google.com/gce-pd”). Note that the CSI driver name can
+be different and this should be just matter of configuration.
+
+Caveat: On downgrade the CSI Driver must still exist to continue the lifecycle
+of the PV’s and volumes it provisioned, the in-tree driver will not pick up
+where the CSI Driver left off. We could have the in-tree volume pick up where
+the CSI driver left off, but in that case the two provisioners race with each
```

Comment:
this seems like a good use case for leader election between in-tree GCE and CSI GCE... only one active at a time, either able to handle storageclass->pv, pv attach/detach, etc